lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -200,16 +201,47 @@ def safe_add_concurrent_partitioned_index(
 
     _validate_index_name!(parent_index)
 
+    # Validate index names upfront to avoid unnecessary recursion
     child_tables_with_metadata = _partitions_for_table(parent_schema, parent_table).map do |child_schema, child_table|
-      raise PgHaMigrations::InvalidMigrationError, "Partitioned table #{parent_table.inspect} contains sub-partitions" if _partitioned_table?(child_schema, child_table)
-
       child_index = "#{child_table}_#{name_suffix}"
 
       _validate_index_name!(child_index)
 
       [child_schema, child_table, child_index]
     end
 
+    # When sub-partitioning detected:
+    #   - recursively create indexes when recurse is true
+    #   - do nothing when if_not_exists is true and index already valid
+    #   - otherwise raise helpful error
+    child_tables_with_metadata.each do |child_schema, child_table, child_index|
+      next unless _partitioned_table?(child_schema, child_table)
+
+      if recurse
+        safe_add_concurrent_partitioned_index(
+          "#{connection.quote_schema_name(child_schema)}.#{child_table}",
+          columns,
+          name_suffix: name_suffix,
+          if_not_exists: if_not_exists,
+          recurse: recurse,
+          using: using,
+          unique: unique,
+          where: where,
+        )
+      elsif !if_not_exists
+        message = "Partitioned table #{parent_table.inspect} contains sub-partitions. " \
+          "If sub-partition indexes have already been created, execute this method with 'if_not_exists: true'. " \
+          "If sub-partition indexes have not been created, consider executing this method with 'recurse: true'."
+
+        raise PgHaMigrations::InvalidMigrationError, message
```

Comment:
This is taking a hint (hehe) from Postgres's error style, but:
```suggestion
        raise PgHaMigrations::InvalidMigrationError, <<~EOS
          Partitioned table #{parent_table.inspect} contains sub-partitions.
          Hint: If sub-partition indexes have already been created, execute this method with 'if_not_exists: true'.
          Hint: If sub-partition indexes have not been created, consider executing this method with 'recurse: true'.
        EOS
```